packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/Notifiable.kt

```diff
@@ -15,38 +15,118 @@
+ * into [C]-change events.
+ */
+public interface Notifiable<T : CoreNotifiable<T, C>, C> {
+
+    /**
+     * Should return the live reference in [liveRealm] that the [SuspendableNotifier] will register
+     * notifications for with Core, or `null` if the entity has been deleted.
+     */
+    public fun coreObservable(liveRealm: LiveRealm): CoreNotifiable<T, C>?
+
+    /**
+     * The [ChangeFlow] responsible for emitting [SuspendableNotifier] events as appropriate
+     * [C]-change events.
+     */
+    public fun changeFlow(scope: ProducerScope<C>): ChangeFlow<T, C>
 }
 
-// TODO Public due to being a transitive dependency to Observable
-public interface Thawable<T> {
-    public fun thaw(liveRealm: RealmReference): T?
+/**
+ * A _change flow_ is responsible for converting the [SuspendableNotifier] events into
+ * corresponding [C]-change events and emit them in the [producerScope] and close the scope when
+ * the monitored entity is deleted.
+ */
+public abstract class ChangeFlow<T, C>(private val producerScope: ProducerScope<C>) {
+
+    private var previousElement: T? = null
+
+    /**
+     * Converts the given [SuspendableNotifier] event into a C-change event, emit it and potentially
+     * close the [producerScope] if the monitored event was deleted (`frozenRef == null`).
+     *
+     * The default implementation will emit a C-event based on the following rules:
+     * - If `frozenRef == null` the result of calling [delete] will be emitted and the
+     *   [producerScope] will be closed.
+     * - If `frozenRef != null` and `previousElement == null` the result of calling [initial] will
+     *   be emitted.
+     * - Otherwise the result of calling [update] will be emitted.
+     *
+     * @param frozenRef a frozen reference of the original entity, or `null` if the entity is no
+     * longer present in the [SuspendableNotifier]'s live realm.
+     * @param change the core change, or `null` if this is the initial event issued by the
+     * [SuspendableNotifier] at the point of callback registration.
+     */
+    internal fun emit(frozenRef: T?, change: RealmChangesPointer? = null) {
+        val event = if (frozenRef != null) {
+            if (previousElement == null) {
+                initial(frozenRef)
+            } else {
+                change?.let { update(frozenRef, it) }
+                    ?: sdkError("We should never receive change callbacks for non-null (deleted) entities without an actual change object")
+            }
+        } else {
+            delete()
+        }
+        previousElement = frozenRef
+        event?.let { producerScope.trySendWithBufferOverflowCheck(it) }
+        if (frozenRef == null) {
+            producerScope.close()
+        }
+    }
+
+    internal abstract fun initial(frozenRef: T): C
+    internal abstract fun update(frozenRef: T, change: RealmChangesPointer): C?
+    internal abstract fun delete(): C
 }
 
 // TODO Why is this flowable here?
```

Comment:
This should be moved to a different package?

Reply:
Which package should i be in? `io.realm.kotlin.internal` seems quite appropriate to me ... but I have moved it to a separate file. Don't know if that changes anything though 🤷